### PR TITLE
WS-12: Minimal viable Create user function

### PR DESCRIPTION
**Pull request by DishaDarpan**
Created a form to create new user. It includes password validation and non-empty validation for other fields. Made a new ticket for css of this page. You need to be logged in as a user to test this function.

### CODE REVIEW

File: frontend/src/components/create-user-page/CreateUser.tsx [hunk 1]
```diff
@@ -1,3 +1,118 @@
...
+  const [formData, setFormData] = useState({
+    name: "",
+    username: "",
+    email: "",
+    password: "",
+  });
+  const validatePassword = (passwordToCheck: string) => {
+    const upper = /(.*[A-Z].*)/;
+    const lower = /(.*[a-z].*)/;
+    const number = /[0-9]/;
```
Comment (TimLeach635) [inline frontend/src/components/create-user-page/CreateUser.tsx:16]: Regex has a handy little single character test thing for this exact purpose:
```suggestion
    const number = /\d/;
```

File: frontend/src/components/create-user-page/CreateUser.tsx [hunk 2]
```diff
@@ -1,3 +1,118 @@
...
+      alert("Please make sure Password Includes a Digit");
+      return false;
+    }
+
+    if (!special.test(passwordToCheck)) {
+      alert(
+        "Please make sure Password Includes a special character from -+_!@#$%^&*.,?"
+      );
+      return false;
+    }
```
Comment (TimLeach635) [inline frontend/src/components/create-user-page/CreateUser.tsx:44]: Using `alert`s is fine for now, but isn't the best UX (pop-up windows are very jarring!)

In your follow-up styling ticket, could you also make sure to replace these with, for example, red text that appears below the fields that aren't correct?

File: frontend/src/components/create-user-page/CreateUser.tsx [hunk 3]
```diff
@@ -1,3 +1,118 @@
...
+
+  const handleSubmit = (e: FormEvent<HTMLFormElement>) => {
+    e.preventDefault();
+    if (
+      formData.name === "" ||
+      formData.username === "" ||
+      formData.email === "" ||
+      formData.password === ""
+    ) {
+      alert("All fields required!");
```
Comment (TimLeach635) [inline frontend/src/components/create-user-page/CreateUser.tsx:55]: As above, avoid alerts - prefer to add text below the specific missing field that says "this field is required" (but, again, you can save that for the follow-up styling ticket)

File: frontend/src/components/create-user-page/CreateUser.tsx [hunk 4]
```diff
@@ -1,3 +1,118 @@
...
+    } else {
+      fetch("https://localhost:5001/users/create", {
+        method: "POST",
+        headers: {
+          // 'Accept': 'application/json',
```
Comment (TimLeach635) [inline frontend/src/components/create-user-page/CreateUser.tsx:60]: Please remove unused comments:
```suggestion
```

File: frontend/src/components/create-user-page/CreateUser.tsx [hunk 5]
```diff
@@ -1,3 +1,118 @@
...
+          "Content-Type": "application/json",
+          Authorization: `Basic ${btoa(
+            `${loginContext.username}:${loginContext.password}`
+          )}`,
+        },
+        body: JSON.stringify(formData),
+      }).then((res) => res.json());
+      //.then(() => setRedirectTarget("/users"));
```
Comment (TimLeach635) [inline frontend/src/components/create-user-page/CreateUser.tsx:68]: Comment
```suggestion
```

File: frontend/src/components/create-user-page/CreateUser.tsx [hunk 6]
```diff
@@ -1,3 +1,118 @@
...
+    }
+  };
+
+  return (
+    <div>
+      <form onSubmit={handleSubmit}>
+        <label>
+          Name :
```
Comment (TimLeach635) [inline frontend/src/components/create-user-page/CreateUser.tsx:76]: Prefer no space before colon (change others in file as well!)
```suggestion
          Name:
```

File: frontend/src/components/create-user-page/CreateUser.tsx [hunk 7]
```diff
@@ -1,3 +1,118 @@
...
+          <input
+            type="text"
+            name="Name"
+            onChange={(e) => setFormData({ ...formData, name: e.target.value })}
```
Comment (TimLeach635) [inline frontend/src/components/create-user-page/CreateUser.tsx:80]: These will be able to be simplified once you switch over from using an object as state (discussed above):
```suggestion
            onChange={e => setFormName(e.target.value)}
```

File: frontend/src/components/create-user-page/CreateUser.tsx [hunk 8]
```diff
@@ -1,3 +1,118 @@
...
+          <input
+            type="password"
+            name="Password"
+            onChange={(e) =>
+              setFormData({ ...formData, password: e.target.value })
+            }
+            onBlur={(e) => validatePassword(e.target.value)}
+          />
+        </label>
+        <button>Create User</button>
```
Comment (TimLeach635) [inline frontend/src/components/create-user-page/CreateUser.tsx:114]: In forms, I prefer the submit button to be an `<input type="submit">`, it feels a little more semantic:
```suggestion
        <input type="submit" value="Create User" />
```

**Review by TimLeach635 — CHANGES_REQUESTED** [review]
Some changes!

File: frontend/src/components/create-user-page/CreateUser.tsx [hunk 9]
```diff
@@ -1,3 +1,118 @@
...
-export const CreateUser: React.FunctionComponent = () => <></>;
+export const CreateUser: React.FunctionComponent = () => {
+  const loginContext = useContext(LoginContext);
+
+  const [formData, setFormData] = useState({
+    name: "",
+    username: "",
+    email: "",
+    password: "",
+  });
```
[the remark this answers is not in the document]
Comment (DishaDarpan) [reply frontend/src/components/create-user-page/CreateUser.tsx:12]: Done

File: frontend/src/components/create-user-page/CreateUser.tsx (hunk 3 above)
Comment (DishaDarpan) [reply frontend/src/components/create-user-page/CreateUser.tsx:55]: So, for this do we need to use useState and set some text for the required field and style it?

File: frontend/src/components/create-user-page/CreateUser.tsx (hunk 2 above)
Comment (DishaDarpan) [reply frontend/src/components/create-user-page/CreateUser.tsx:44]: Ok. But may need some help with this.

Comment (DishaDarpan) [reply frontend/src/components/create-user-page/CreateUser.tsx:43]: We will resolve this when we do styling ticket 60

File: frontend/src/components/create-user-page/CreateUser.tsx (hunk 3 above)
Comment (DishaDarpan) [reply frontend/src/components/create-user-page/CreateUser.tsx:54]: We will resolve this when we do styling ticket 16.

**Review by TimLeach635 — APPROVED** [review]

**Pull request merged**